# Chapter 4. Working with Text

Companion notebook for *Python from Zero*, Chapter 4. Run the cells from
top to bottom with **Shift + Enter**. Some cells show an error on purpose:
reading those messages is part of the lesson.

## 4.1 Three kinds of quotes

In [1]:
print("Sunny Paws Animal Rescue")
print('Captain is the office cat')

Sunny Paws Animal Rescue
Captain is the office cat


In [2]:
print("Captain's favorite spot is the printer")
print('Rosa said "Welcome aboard!" on your first day')

Captain's favorite spot is the printer
Rosa said "Welcome aboard!" on your first day


In [3]:
print('Captain's favorite spot is the printer')

SyntaxError: unterminated string literal (detected at line 1) (2872311.py, line 1)

In [4]:
sign = """Sunny Paws Animal Rescue
Open Tuesday to Sunday, 10:00 to 17:00
Closed Mondays"""
print(sign)

Sunny Paws Animal Rescue
Open Tuesday to Sunday, 10:00 to 17:00
Closed Mondays


## 4.2 Escape characters

In [5]:
print("Name:\tCaptain\nAge:\t12\nJob:\tOffice cat")

Name:	Captain
Age:	12
Job:	Office cat


In [6]:
print("Rosa said \"Welcome aboard!\" on your first day")
print('Captain\'s favorite spot is the printer')
print("Photos are in C:\\Shelter\\photos")

Rosa said "Welcome aboard!" on your first day
Captain's favorite spot is the printer
Photos are in C:\Shelter\photos


## 4.3 How long is a string? len()

In [7]:
name = "Captain"
print(len(name))
print(len("Sunny Paws Animal Rescue"))
print(len(""))

7
24
0


## 4.4 Indexing: one character at a time

In [8]:
name = "Captain"
print(name[0])
print(name[1])
print(name[6])

C
a
n


In [9]:
print(name[-1])
print(name[-2])
print(name[-7])

n
i
C


In [10]:
print(name[7])

IndexError: string index out of range

## 4.5 Slicing: taking a piece

In [11]:
animal_id = "SP-014"
print(animal_id[0:2])
print(animal_id[3:6])

SP
014


In [12]:
print(animal_id[:2])
print(animal_id[3:])
print(animal_id[-3:])
print(animal_id[:])

SP
014
014
SP-014


In [13]:
arrived = "2026-02-18"
year = arrived[:4]
month = arrived[5:7]
day = arrived[8:]
print("Year: " + year)
print("Month: " + month)
print("Day: " + day)

Year: 2026
Month: 02
Day: 18


In [14]:
print(animal_id[3:100])
print(animal_id[10:20])
print("[" + animal_id[10:20] + "]")

014

[]


In [15]:
word = "Captain"
print(word[::2])
print(word[::-1])

Cpan
niatpaC


## 4.6 Strings cannot be changed

In [16]:
name = "Captain"
name[0] = "K"

TypeError: 'str' object does not support item assignment

In [17]:
new_name = "K" + name[1:]
print(new_name)
print(name)

Kaptain
Captain


## 4.7 String methods

In [18]:
visitor = "priya shah"
print(visitor.upper())
print(visitor.title())
print("PRIYA SHAH".lower())

PRIYA SHAH
Priya Shah
priya shah


In [19]:
print(visitor)
visitor = visitor.title()
print(visitor)

priya shah
Priya Shah


In [20]:
raw = "   sam whitfield  "
print("[" + raw + "]")
print("[" + raw.strip() + "]")
print("[" + raw.lstrip() + "]")
print("[" + raw.rstrip() + "]")

[   sam whitfield  ]
[sam whitfield]
[sam whitfield  ]
[   sam whitfield]


In [21]:
clean = raw.strip().title()
print(clean)

Sam Whitfield


In [22]:
arrived = "2026-02-18"
print(arrived.replace("-", "/"))
note = "Walk Peanut. Feed Peanut. Brush Peanut."
print(note.replace("Peanut", "Scout"))
print(note.replace("Rocket", "Scout"))

2026/02/18
Walk Scout. Feed Scout. Brush Scout.
Walk Peanut. Feed Peanut. Brush Peanut.


In [23]:
animal_id = "SP-014"
print(animal_id.startswith("SP-"))
print(animal_id.startswith("sp-"))
file_name = "feeding_chart.txt"
print(file_name.endswith(".txt"))
print(file_name.endswith(".csv"))

True
False
True
False


In [24]:
line = "2026-06-03 14:28 | Zoe Martin | asked about Truffle"
print(line.find("|"))
print(line.find("Truffle"))
print(line.find("Mochi"))
print(line.count("|"))
print(line.count("a"))

17
44
-1
2
3


In [25]:
print("Truffle" in line)
print("Mochi" in line)
print("truffle" in line)
print("truffle" in line.lower())

True
False
False
True


In [26]:
print(line.split("|"))
print("Peanut,Dog,Labrador mix".split(","))
print("Sunny Paws Animal Rescue".split())

['2026-06-03 14:28 ', ' Zoe Martin ', ' asked about Truffle']
['Peanut', 'Dog', 'Labrador mix']
['Sunny', 'Paws', 'Animal', 'Rescue']


## 4.8 f-strings

In [27]:
name = "Peanut"
species = "Dog"
age = 4.5
print(f"{name} is a {age}-year-old {species}.")

Peanut is a 4.5-year-old Dog.


In [28]:
print("Age: " + age)

TypeError: can only concatenate str (not "float") to str

In [29]:
print(f"Age: {age}")
print(f"{name.upper()} has {len(name)} letters.")
print(f"In months, that is {age * 12} months.")

Age: 4.5
PEANUT has 6 letters.
In months, that is 54.0 months.


In [30]:
fee = 150
print(f"Fee: ${fee}")
print(f"Fee: ${fee:.2f}")
weight = 28.6
per_day = weight * 15
print(f"Food per day: {per_day} g")
print(f"Food per day: {per_day:.0f} g")

Fee: $150
Fee: $150.00
Food per day: 429.0 g
Food per day: 429 g


In [31]:
total_donations = 7805
print(f"Donations so far: ${total_donations:,}")
print(f"Donations so far: ${total_donations:,.2f}")
print(f"Average gift: ${total_donations / 130:.2f}")

Donations so far: $7,805
Donations so far: $7,805.00
Average gift: $60.04


In [32]:
name = "Kiwi"
print(f"[{name:<10}]")
print(f"[{name:>10}]")
print(f"[{name:^10}]")

[Kiwi      ]
[      Kiwi]
[   Kiwi   ]


In [33]:
dog_fee = 150
cat_fee = 90
rabbit_fee = 45
guinea_pig_fee = 25
print(f"{'Species':<12}{'Fee ($)':>10}")
print("-" * 22)
print(f"{'Dog':<12}{dog_fee:>10.2f}")
print(f"{'Cat':<12}{cat_fee:>10.2f}")
print(f"{'Rabbit':<12}{rabbit_fee:>10.2f}")
print(f"{'Guinea pig':<12}{guinea_pig_fee:>10.2f}")

Species        Fee ($)
----------------------
Dog             150.00
Cat              90.00
Rabbit           45.00
Guinea pig       25.00


In [34]:
print(f"{'Sunny Paws':*^30}")
print(f"{' Adoption Day ':=^30}")

**********Sunny Paws**********
======== Adoption Day ========


## 4.9 Project: a name tag and a kennel card

In [35]:
name = "Toast"
species = "Cat"
breed = "Maine Coon mix"
age = 3.5

print("+" + "-" * 28 + "+")
print(f"|{name.upper():^28}|")
print(f"|{species + ', ' + breed:^28}|")
print(f"|{'Age: ' + str(age) + ' years':^28}|")
print("+" + "-" * 28 + "+")

+----------------------------+
|           TOAST            |
|    Cat, Maine Coon mix     |
|       Age: 3.5 years       |
+----------------------------+


In [36]:
record = "SP-014,Kiwi,Dog,Beagle,F,3,12.2,2026-02-18,Available"
kennel = 7

animal_id = record[:6]
name = "Kiwi"
breed = "Beagle"
weight = 12.2
arrived = record[-20:-10]

print("=" * 32)
print(f" KENNEL {kennel:<10}{animal_id:>13}")
print("=" * 32)
print(f" {'Name:':<10}{name}")
print(f" {'Breed:':<10}{breed}")
print(f" {'Weight:':<10}{weight:.1f} kg")
print(f" {'Arrived:':<10}{arrived}")
print(f" {'Food:':<10}{weight * 15:.0f} g dry food a day")
print("=" * 32)

 KENNEL 7                SP-014
 Name:     Kiwi
 Breed:    Beagle
 Weight:   12.2 kg
 Arrived:  2026-02-18
 Food:     183 g dry food a day


In [37]:
line = "2026-06-03 13:29 |  PRIYA SHAH |  asked about Rosie"
first_bar = line.find("|")
second_bar = line.find("|", first_bar + 1)
print(first_bar)
print(second_bar)

17
31


In [38]:
date = line[:10]
visitor = line[first_bar + 1:second_bar].strip().title()
reason = line[second_bar + 1:].strip()
print(f"{date}  {visitor:<14}{reason}")

2026-06-03  Priya Shah    asked about Rosie


## 4.10 Common mistakes

In [39]:
name = "luna"
name.title()
print(name)

luna


In [40]:
name = "Luna"
print("{name} is ready for adoption")
print(f"{name} is ready for adoption")

{name} is ready for adoption
Luna is ready for adoption


In [41]:
print(name.upper)

<built-in method upper of str object at 0x7f04c200dd10>


## Exercises

1. Use one `print()` with `\t` and `\n` to show three facts about Captain
   (name, age 12, color ginger), each on its own line with a tab after the label.
2. From `animal_id = "SP-037"` print the number part `037`; from
   `arrived = "2026-05-01"` print the month `05`.
3. Clean `messy = "   jUNIPER  "` into `Juniper` and print it with its length.
4. From `line = "2026-06-07 12:02 |  PRIYA SHAH |  asked about Pumpkin"` print
   the date, the time, whether the line mentions Pumpkin, and how many `|` it has.
5. With `total = 7805` and `count = 130`, print
   `Total: $7,805.00 from 130 donations (average $60.04)`.
6. Print the four adoption fees as an aligned table, with the title
   `Sunny Paws Adoption Fees` centered above it and padded with `=` signs,
   and a dollar sign before each fee.
7. Make a boxed name tag for Captain, 24 characters wide inside the walls:
   `CAPTAIN`, `Ginger cat, age 12`, `Office cat`.
8. Challenge: pull the visitor's name and reason out of
   `"2026-06-02 14:53|  sam whitfield|returned Shadow from foster"`
   and print `Sam Whitfield: returned Shadow from foster`. Check that your
   code also works on `"2026-06-13 11:00 |    sam whitfield |  asked about Toast"`.